src/m_array.c

```diff
@@ -0,0 +1,94 @@
+//
+// Copyright(C) 2024 Roman Fomin
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#include <stddef.h>
+#include <stdlib.h>
+#include "i_system.h"
+
+#define ARRAY_INIT_CAPACITY 8
+
+struct buffer_s
+{
+    unsigned int capacity;
+    unsigned int size;
+    char buffer[];
+};
+
+#define ARR_PTR(v) \
+    ((struct buffer_s *)((char *)(v) - offsetof(struct buffer_s, buffer)))
```

Comment:
The trick is to use the `offsetof` macro from `stddef.h` and store the capacity and size in front of the array buffer.

Reply:
Never heard of this before, but it's brilliant!